holidays/countries/thailand.py

```diff
@@ -12,156 +12,633 @@
-                while in_lieu.year == year and in_lieu in self:
-                    in_lieu += td(days=+1)
-
-                add_holiday(in_lieu, f"{holiday_name} (in lieu)")
+        # Due to Thai Calendar Migration, this is capped off at 1941
+        # But certain holidays were implemented before 1941
+        if year <= 1940:
+            return
+
+        def _add_with_observed(dt, holiday_name) -> None:
+            # Only add if current year and year is 1941 (B.E. 2484) or later
+            # This is here as a stub for future islamic consecutive holidays
+            # Which can stradle across gregorian years in southern region
+            if dt.year == year:
+                self[dt] = holiday_name
+
+            """
+            !!! If Public Holiday falls on weekends, (in lieu) on workday !!!
+            Despite the wording, this usually only applies to Monday only for
+            holidays, consecutive holidays all have their own special in lieu
+            declared separately.
+
+            Data from 1992-1994 and 1998-2000 are declared discretely in
+            special_holidays declarations above.
+
+            Applied Automatically for Monday if on Weekends: 1961-1973
+             **NOTE: No New Year's Eve (in lieu) for this period
+            No In Lieu days available: 1974-1988
+            Case-by-Case application for Workday if on Weekends: 1989-1994
+            Applied Automatically for Workday if on Weekends: 1995-1997
+            Case-by-Case application for Workday if on Weekends: 1998-2000
+            Applied Automatically for Workday if on Weekends: 2001-Present
+            """
+            if 1961 <= year <= 1973 or 1995 <= year <= 1997 or year >= 2001:
+                if self.observed and self._is_weekend(dt):
+                    in_lieu = dt + td(days=+2 if dt.weekday() == SAT else +1)
+                    self[in_lieu] = f"{holiday_name} (in lieu)"
 
         super()._populate(year)
 
-        # New Year's Day.
-        add_holiday(date(year, 1, 1), "New Year's Day")
-
-        # Magha Pujab.
-        # TODO(ark): research Buddhist calendar options.
-        magha_puja = {
-            2016: (FEB, 22),
-            2017: (FEB, 11),
-            2018: (MAR, 1),
-            2019: (FEB, 19),
-            2020: (FEB, 26),
-            2021: (FEB, 26),
-            2022: (FEB, 16),
-            2023: (MAR, 6),
-        }
-        if year in magha_puja:
-            add_holiday(date(year, *magha_puja[year]), "Makha Bucha")
-
-        # Chakri Memorial Day.
-        add_holiday(date(year, APR, 6), "Chakri Memorial Day")
-
-        # Songkran Festival.
+        ########################
+        #
+        # FIXED DATED HOLIDAYS
+        #
+        ########################
+
+        # !!! New Year's Day !!!
+        # วันขึ้นปีใหม่
+        # Status: In-Use
+        # Starts in the present form in 1941 (B.E. 2484)
+        # [TODO] Add check for 1941 if we support earlier dates.
+
```

Comment:
The vertical indents after the descriptions can be removed as those look like single logical blocks.